### DISCUSSION TITLE
ztunnel random errors for OUTBOUND(external) traffic

### DISCUSSION THREAD

**Discussion by sachintiptur** [istio/istio#51983] [Q&A]
We have installed Istio ambient v1.22.2 in our cluster and enabled it for our monitoring services.
Below errors are randomly seen in ztunnel. These outbound traffic errors are seen only for the traffic going out of the pod(in mesh) to external service. In this case, it is while sending alerts from Prometheus-msteams service to external client MS teams channel. But we see some alerts are reaching MS teams and this errors are random. Can you please shed some light on these? whether the errors are real or what can attribute to these errors? Any configuration or resources need to be defined for this use case? Your response would be much appreciated.

```
2024-07-10T09:27:55.848249Z     error   access  connection complete     src.addr=10.233.121.171:58050 src.workload="prometheus-msteams-59744677f7-8fcsd" src.namespace="monitoring" dst.addr=xx.xx.176.2:443 direction="outbound" bytes_sent=0 bytes_recv=0 duration="10001ms" error="io error: deadline has elapsed"
2024-07-10T09:28:10.350738Z     error   access  connection complete     src.addr=10.233.121.171:45806 src.workload="prometheus-msteams-59744677f7-8fcsd" src.namespace="monitoring" dst.addr=xx.xy.201.114:443 direction="outbound" bytes_sent=0 bytes_recv=0 duration="10001ms" error="io error: deadline has elapsed"
```

**Comment by howardjohn** [comment]
Doe the issue happen for all calls to the external site, or just intermittently? Are there some NetworkPolicies blocking things perhaps?

What k8s platform/CNI are you using?

For this codepath, ztunnel is not doing much really. Just acting as a simple TCP proxy

**Comment by sachintiptur** [comment]
It is happening intermittently, because we are receiving alerts. We are not using any NetworkPolicies. 
Ours is a bare metal k8s v1.29.5 cluster with Calico v3.27.3. 
Oh ok, in that case, can you please point out what to look for to debug this further? like any other logs or iptables/routing tables?

**Reply by howardjohn** [reply]
I've noticed every related issue like this is running Calico, so I am going to do some more investigation on calico. https://github.com/cilium/pwru or similar could maybe see what is dropping packets. My wild guess is something about rp_filter but not quite sure

**Reply by sachintiptur** [reply]
```
./pwru 'dst host 10.233.83.17 and dst port 20257'

SKB                CPU PROCESS          FUNC
0xffff9600c2a2b400 1   ~metheus:2399658 ip_local_out
0xffff9600c2a2b400 1   ~metheus:2399658 __ip_local_out
0xffff9600c2a2b400 1   ~metheus:2399658 nf_hook_slow
0xffff9600c2a2b400 1   ~metheus:2399658 skb_ensure_writable
0xffff9600c2a2b400 1   ~metheus:2399658 skb_ensure_writable
0xffff9600c2a2b400 1   ~metheus:2399658 ip_local_out
0xffff9600c2a2b400 1   ~metheus:2399658 __ip_local_out
0xffff9600c2a2b400 1   ~metheus:2399658 nf_hook_slow
0xffff9600c2a2b400 1   ~metheus:2399658 ip_route_me_harder
0xffff9600c2a2b400 1   ~metheus:2399658 __xfrm_decode_session
0xffff9600c2a2b400 1   ~metheus:2399658 decode_session4
0xffff9600c2a2b400 1   ~metheus:2399658 security_xfrm_decode_session
0xffff9600c2a2b400 1   ~metheus:2399658 skb_ensure_writable
0xffff9600c2a2b400 1   ~metheus:2399658 skb_ensure_writable
0xffff9600c2a2ac00 1   ~metheus:2399658 ip_local_out
0xffff9600c2a2ac00 1   ~metheus:2399658 __ip_local_out
0xffff9600c2a2ac00 1   ~metheus:2399658 nf_hook_slow
0xffff9600c2a2ac00 1   ~metheus:2399658 ip_route_me_harder
0xffff9600c2a2ac00 1   ~metheus:2399658 __xfrm_decode_session
0xffff9600c2a2ac00 1   ~metheus:2399658 decode_session4
0xffff9600c2a2ac00 1   ~metheus:2399658 security_xfrm_decode_session
0xffff9600c2a2ac00 1   ~metheus:2399658 skb_ensure_writable
0xffff9600c2a2ac00 1   ~metheus:2399658 skb_ensure_writable

```

**Reply by sachintiptur** [reply]
This is the trace I got for below error in ztunnel. this time it was for in-cluster traffic
`2024-07-18T09:00:44.150138Z     error   access  connection complete     src.addr=xx.yy.86.218:51764 src.workload="prometheus-kube-prometheus-stack-prometheus-0" src.namespace="monitoring" src.identity="spiffe://cluster.local/ns/monitoring/sa/kube-prometheus-stack-prometheus" dst.addr=10.233.83.17:15008 dst.hbone_addr=xx.yy.83.17:20257 dst.workload="node-problem-detector-4tjn7" dst.namespace="npd" dst.identity="spiffe://cluster.local/ns/npd/sa/node-problem-detector" direction="outbound" bytes_sent=0 bytes_recv=0 duration="0ms" error="io error: Connection refused (os error 111)"`
